core/server/src/main/java/alluxio/master/file/meta/Inode.java

```diff
@@ -277,24 +292,93 @@ public T setOwner(String owner) {
   protected abstract T getThis();
 
   /**
-   * Acquires the read lock for this inode.
+   * Obtains a read lock on the inode. This should only be used when the inode is accessed by ID
+   * and not by path or parent.
    */
   public void lockRead() {
     mLock.readLock().lock();
   }
 
   /**
-   * Releases the read lock for this inode.
+   * Obtains a read lock on the inode. Afterward, checks the inode state to ensure the full inode
+   * path is consistent with what the caller is expecting. If the state is inconsistent, an
+   * exception will be thrown and the lock will be released.
+   *
+   * @param parent the expected parent inode
+   * @param name the expected name of the inode to be locked
+   * @throws InvalidPathException if the parent and/or name is not as expected
+   */
+  public void lockReadAndCheckFullPath(Inode parent, String name) throws InvalidPathException {
+    lockReadAndCheckParent(parent);
+    if (!mName.equals(name)) {
+      mLock.readLock().unlock();
+      throw new InvalidPathException("Path is no longer valid, possibly due to a concurrent move.");
+    }
+  }
+
+  /**
+   * Obtains a read lock on the inode. Afterward, checks the inode state to ensure the parent is
+   * consistent with what the caller is expecting. If the state is inconsistent, an exception
+   * will be thrown and the lock will be released.
+   *
+   * @param parent the expected parent inode
+   * @throws InvalidPathException if the parent is not as expected
    */
-  public void unlockRead() {
-    mLock.readLock().unlock();
+  public void lockReadAndCheckParent(Inode parent) throws InvalidPathException {
+    mLock.readLock().lock();
+    if (mParentId != InodeTree.NO_PARENT && mParentId != parent.getId()) {
+      mLock.readLock().unlock();
+      throw new InvalidPathException("Path is no longer valid, possibly due to a concurrent move.");
+    }
   }
 
   /**
-   * Acquires the write lock for this inode.
+   * Obtains a write lock on the inode. This should only be used when the inode is accessed by ID
+   * and not by path or parent.
    */
   public void lockWrite() {
+    mLock.readLock().lock();
```

Comment:
did our unit / integration tests not catch this? if not, please add a test that would catch this next time